res/res_crypto.c

```diff
@@ -797,6 +792,383 @@ int AST_OPTIONAL_API_NAME(ast_aes_decrypt)(const unsigned char *in, unsigned cha
 	return res;
 }
 
+/*
+ * BEGIN OPENSSL HELPERS
+ */
+
+static char *openssl_error(void)
+{
+	BIO *bio = BIO_new(BIO_s_mem());
+	char *buf = NULL;
+	char *text;
+	int len;
+
+	ERR_print_errors(bio);
+	len = BIO_get_mem_data(bio, &buf);
+	text = ast_malloc(len + 1);
+	if (!text) {
+		BIO_free(bio);
+		return NULL;
+	}
+	memcpy(text, buf, len);
+	text[len] = '\0';
+	BIO_free(bio);
+	return text;
+}
```

Comment:
I'm certainly biased but I'm partial to this implementation:

https://github.com/asterisk/asterisk/blob/87e053f6d1809366704de2e4a8a2338cd621d35c/main/tcptls.c#L111-L130

Reply:
we have openssl stuff spread out in too many places :(.

it'd be better if you could pass it a message as well.  maybe i can just make it a va_arg.

Reply:
made a new function: ast_log_openssl.